In [1]:
pip install scikit-learn

Could not fetch URL https://pypi.org/simple/pip/: There was a problem confirming the ssl certificate: HTTPSConnectionPool(host='pypi.org', port=443): Max retries exceeded with url: /simple/pip/ (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)'))) - skipping
Note: you may need to restart the kernel to use updated packages.


In [2]:
# Import the modules
import pandas as pd
import hvplot.pandas
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [7]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_market_data = pd.read_csv("lending_data.csv", 
                             index_col="loan_size")


# Review the DataFrame
df_market_data.head(10)

interest_rate  borrower_income  debt_to_income  num_of_accounts  \
loan_size                                                                    
10700.0            7.672            52800        0.431818                5   
8400.0             6.692            43600        0.311927                3   
9000.0             6.963            46100        0.349241                3   
10700.0            7.664            52700        0.430740                5   
10800.0            7.698            53000        0.433962                5   
10100.0            7.438            50600        0.407115                4   
10300.0            7.490            51100        0.412916                4   
8800.0             6.857            45100        0.334812                3   
9300.0             7.096            47400        0.367089                3   
9700.0             7.248            48800        0.385246                4   

           derogatory_marks  total_debt  loan_status  
loan_size                                             
10700.0                   1       22800            0  
8400.0                    0       13600            0  
9000.0                    0       16100            0  
10700.0                   1       22700            0  
10800.0                   1       23000            0  
10100.0                   1       20600            0  
10300.0                   1       21100            0  
8800.0                    0       15100            0  
9300.0                    0       17400            0  
9700.0                    0       18800            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [8]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_market_data["loan_status"]

# Separate the X variable, the features
X = df_market_data.drop(columns=["loan_status"])

In [9]:
# Review the y variable Series
print(y.value_counts())

loan_status
0    75036
1     2500
Name: count, dtype: int64


In [10]:
print(y.head())

loan_size
10700.0    0
8400.0     0
9000.0     0
10700.0    0
10800.0    0
Name: loan_status, dtype: int64


In [12]:
# Review the X variable DataFrame
print(X.head())

           interest_rate  borrower_income  debt_to_income  num_of_accounts  \
loan_size                                                                    
10700.0            7.672            52800        0.431818                5   
8400.0             6.692            43600        0.311927                3   
9000.0             6.963            46100        0.349241                3   
10700.0            7.664            52700        0.430740                5   
10800.0            7.698            53000        0.433962                5   

           derogatory_marks  total_debt  
loan_size                                
10700.0                   1       22800  
8400.0                    0       13600  
9000.0                    0       16100  
10700.0                   1       22700  
10800.0                   1       23000  


### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [13]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [14]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression


# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [15]:
# Make a prediction using the testing data
predictions = logistic_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Generate a confusion matrix for the model
cm = confusion_matrix(y_test, predictions)

In [18]:
# Print the classification report for the model
print("Confusion Matrix:")
print(cm)
# Print the classification report
print("\nClassification Report:")
print(classification_report(y_test, predictions))

Confusion Matrix:
[[18656   109]
 [   33   586]]

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.95      0.89       619

    accuracy                           0.99     19384
   macro avg       0.92      0.97      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** True Negatives (TN): 18656 (healthy loans correctly predicted)
False Positives (FP): 109 (healthy loans incorrectly predicted as high-risk)
False Negatives (FN): 33 (high-risk loans incorrectly predicted as healthy)
True Positives (TP): 586 (high-risk loans correctly predicted)
Classification Report:
Precision:

For label 0: 1.00 (perfect precision, indicating that all predicted healthy loans were indeed healthy)
For label 1: 0.84 (indicating that 84% of the predicted high-risk loans were actually high-risk)
Recall:

For label 0: 0.99 (indicating that 99% of actual healthy loans were correctly identified)
For label 1: 0.95 (indicating that 95% of actual high-risk loans were correctly identified)
F1-Score:

For label 0: 1.00 (perfect balance between precision and recall for healthy loans)
For label 1: 0.89 (a good balance for high-risk loans, but slightly lower than for healthy loans)
Overall Accuracy:
The overall accuracy of the model is 0.99, meaning that 99% of the predictions made by the model were correct.
Summary:
The logistic regression model performs exceptionally well in predicting healthy loans (0), with perfect precision and a high recall.
It also performs well for high-risk loans (1), with a precision of 0.84 and a recall of 0.95, indicating that while it may misclassify some high-risk loans as healthy, it correctly identifies a large majority of them.
The model is robust, as evidenced by the high overall accuracy and strong performance metrics for both classes.
In conclusion, the model is effective at distinguishing between healthy and high-risk loans, with a slight trade-off in precision for the high-risk category. This performance is generally acceptable in many real-world applications, especially considering the high recall for high-risk loans, which is crucial in minimizing financial risk.

---